packages/algob/src/builtin-tasks/compile.ts

```diff
@@ -1,13 +1,112 @@
+import { readdirSync, readFileSync, writeFileSync } from 'fs';
+import * as path from 'path';
+import YAML from "yaml";
+
 import { task } from "../internal/core/config/config-env";
-// import { createClient } from "../lib/driver";
-// import { AlgobRuntimeEnv, TaskArguments } from "../types";
+import { parseAlgorandError } from "../internal/core/errors";
+import { assertDir, ASSETS_DIR, CACHE_DIR } from "../internal/core/project-structure";
+import { createClient } from "../lib/driver";
+import type { AlgobRuntimeEnv, ASCCache, Network } from "../types";
 import { TASK_COMPILE } from "./task-names";
+const murmurhash = require('murmurhash'); // eslint-disable-line @typescript-eslint/no-var-requires
+
+const tealExt = ".teal";
 
 export default function (): void {
-  task(TASK_COMPILE, "Compilation task")
-    .setAction(
-      async () => {
-        console.log("compiling......");
-      }
-    );
+  task(TASK_COMPILE, "Compile all TEAL smart contracts")
+    .addFlag("force", "recompile even if the source file didn't change")
+    .setAction(_compile);
+}
+
+export interface TaskArgs {
+  force: boolean
+}
+
+async function _compile ({ force }: TaskArgs, env: AlgobRuntimeEnv): Promise<void> {
+  const op = new CompileOp(env.network);
+  return await compile(force, op);
+}
+
+export async function compile (force: boolean, op: CompileOp): Promise<void> {
+  await assertDir(CACHE_DIR);
+  const cache = readArtifacts(CACHE_DIR);
+
+  for (const f of readdirSync(ASSETS_DIR)) {
+    if (!f.endsWith(tealExt)) { continue; }
+
+    let c = cache.get(f);
+    const [teal, thash] = readTealAndHash(path.join(ASSETS_DIR, f));
+    if (!force && c !== undefined && c.srcHash === thash) {
+      console.log(`smart-contract source "${f}" didn't change, skipping.`);
+      continue;
+    }
+    console.log("compiling", f);
+    c = await compileAndSave(op, new ASCCachePartial(f, teal, thash));
+    cache.set(f, c);
+    const cacheFilename = path.join(CACHE_DIR, f + ".yaml");
+    op.writeFile(cacheFilename, YAML.stringify(c));
+  }
+}
+
+class CompileOp {
+  algocl: Algodv2;
+
+  constructor (n: Network) {
+    this.algocl = createClient(n);
+  }
+
+  async compile (code: string): Promise<CompileOut> {
+    return await this.algocl.compile(code).do();
+  }
+
+  writeFile (filename: string, content: string): void {
+    writeFileSync(filename, content);
+  }
+}
+
+function readArtifacts (dir: string): Map<string, ASCCache> {
+  const cache = new Map<string, ASCCache>();
+  for (const f of readdirSync(dir)) {
+    const pathCFile = path.join(dir, f);
+    const a = YAML.parse(readFileSync(pathCFile, 'utf8')) as ASCCache;
+    cache.set(f.substring(0, f.length - tealExt.length), a);
+  }
+  return cache;
+}
+
+function readTealAndHash (filename: string): [string, number] {
+  const content = readFileSync(filename, 'utf8');
+  return [content, murmurhash.v3(content)];
+}
+
+class ASCCachePartial {
+  filename: string;
```

Comment:
You have a cache indexed by filename and you have it here again `{filename -> {filename}}`.
The only things needed are hash and maybe a timestamp for further filtering which you don't currently do.

I would construct this data type only after the compilation and only leave the index to have
`{filename -> {hash, timestamp}}`.

Reply:
OK, I can remove the type. I needed it more in the early version of the code.